test/extended/images/hardprune.go

```diff
@@ -475,22 +475,60 @@ func ConfigureRegistry(oc *exutil.CLI, desiredState RegistryConfiguration) error
 
 	err = oc.Run("env").Args(append([]string{"dc/docker-registry"}, envOverrides...)...).Execute()
 	if err != nil {
-		return fmt.Errorf("failed to update registry's environment with %s: %v", &waitForVersion, err)
+		return false, fmt.Errorf("failed to update registry's environment with %s: %v", &waitForVersion, err)
 	}
-	return exutil.WaitForDeploymentConfig(
+	err = exutil.WaitForDeploymentConfig(
 		oc.AdminKubeClient(),
 		oc.AdminAppsClient().Apps(),
 		metav1.NamespaceDefault,
 		"docker-registry",
 		waitForVersion,
 		oc)
+	if err != nil {
+		return false, err
+	}
+	return true, nil
+}
+
+func RedeployRegistry(oc *exutil.CLI) (bool, error) {
+	defer func(ns string) { oc.SetNamespace(ns) }(oc.Namespace())
+
+	oc = oc.SetNamespace(metav1.NamespaceDefault).AsAdmin()
+
+	dc, err := oc.AppsClient().Apps().DeploymentConfigs(metav1.NamespaceDefault).Get("docker-registry", metav1.GetOptions{})
+	if err != nil {
+		return false, err
+	}
+
+	// log docker-registry pod output before re-deploying
+	waitForVersion := dc.Status.LatestVersion + 1
+	if err = LogRegistryPod(oc); err != nil {
+		fmt.Fprintf(g.GinkgoWriter, "failed to log registry pod: %v\n", err)
+	}
+
+	_, err = oc.Run("rollout").Args("latest", "dc/docker-registry", "--again=true").Output()
```

Comment:
@bparees Please review whole PR. The `ConfigureRegistry` reconfigures the registry server, if necessary. This function is waiting for the completion of deploy and does not end before it (`exutil.WaitForDeploymentConfig`).

But observations show that the test is not working only in some cases. These are the cases when several tests are executed in order that the reconfiguration and redeploy does not happen.

I added a forced redeploy if the reconfiguration did not happen. So before each test there is only one registry restart. I use the same method of waiting as the `ConfigureRegistry`.

Reply:
> @bparees Please review whole PR. 

Indeed I did and I understand that intent.  However I'm still not clear on what purpose the "--again=true" serves.

As I recall, when implementing my suggestion to use "oc rollout" instead of update the env variables, you ran into issues where a rollout was already in progress.  It's not obvious to me that anything about this PR has changed to prevent that.  Perhaps now it is being more careful to ensure Redeploy is only invoked when necessary, but I thought it was doing that before as well.

Regardless:

1) if you're confident no rollout will be in progress when Redeploy is invoked, then this is fine, as I said.
2) i'd still like to understand why you're passing "--again=true".  My recollection of the discussion was that you put it in as an attempt to address (1), but it does not address (1).


From our previous discussion (both quotes from you):
```
Now I have an answer for you. Because if you execute the oc rollout latest, you will get: "error: #6 is already in progress (Running)."
```

```
@bparees I added the --again=true option and it seems like everything is working now. That's better ?
```

Reply:
@bparees OK. I will make new research and then will make new PR.